contracts/PetPark.sol

```diff
@@ -1,7 +1,90 @@
-//SPDX-License-Identifier: Unlicense
+// SPDX-License-Identifier: Unlicense
 pragma solidity ^0.8.0;
 
 
 contract PetPark {
 
-}
\ No newline at end of file
+    address owner; 
+
+    enum AnimalType {
+        NONE,
+        Fish,
+        Cat,
+        Dog,
+        Rabbit,
+        Parrot
+    }
+
+    enum Gender {
+        Male,
+        Female
+    }
+
+    struct Borrower {
+        AnimalType animal;
+        bool isBorrowing;
+        Gender gender;
+        uint age;
+    }
+    
+    modifier onlyOwner(){
+        require(msg.sender == owner, "Not an owner");
+        _;
+    }
+   
+    event Added(AnimalType _animalType, uint count);
+    event Borrowed(AnimalType _animalType);
+    event Returned(AnimalType _animalType);
+
+    mapping(AnimalType => uint) animalCount;
+    mapping(address => Borrower) public borrowers;
+
+    constructor(){
+        owner = msg.sender;
+
+    }
+
+    function add(AnimalType _animalType, uint count) public onlyOwner {
+        require(uint(_animalType) > 0, "Invalid animal");
+        animalCount[_animalType] = animalCount[_animalType] + count;
+        emit Added(_animalType, count);
+    }
+
+    function animalCounts(AnimalType _animalType) public view returns (uint) {
+        return animalCount[_animalType];
+    }
+
+    function borrow(uint _age, Gender _gender, AnimalType _animalType) public {
+        require(_age > 0, "Invalid Age");
+        require(uint(_animalType) > 0, "Invalid animal type");
+        require(animalCount[_animalType] > 0, "Selected animal not available");
+        if (borrowers[msg.sender].isBorrowing == true && _age != borrowers[msg.sender].age) {
```

Comment:
Since `borrowers[msg.sender]` is being used multiple times it can be stored in a variable. Later in the course you will learn that it is a gas optimisation technique

Reply:
Just in memory should be fine and you can use that to make all the reads

Reply:
Ok... This is good... I have implemented this w my latest commit 

I noticed that on line 70... I couldnt replace `borrowers[msg.sender]` with the variable `currentUser` that we set up in `memory`. 

Breaks the tests, from what I assume is because the variable is a reference to the spot in memory and can only read, not update? ...

Reply:
You can update it but it will not be persistent. So for example, if you updated the `age` in currentUser, it will not change the age in the actual state